engine/src/main/java/org/hibernate/search/engine/search/predicate/factories/FilterFactoryContext.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.search.predicate.factories;
+
+import java.util.Collection;
+import org.hibernate.search.engine.search.predicate.dsl.FilterPredicateOptionsStep;
+import org.hibernate.search.engine.search.predicate.dsl.SearchPredicateFactory;
+
+/**
+ * The context provided to the filter factory {@link SearchPredicateFactory#def(String)} method.
+ */
+public interface FilterFactoryContext {
+
+	/**
+	 * @return An entry point allowing to declare expectations and retrieve accessors to the bridged POJO type.
+	 * Usual compatibility policies do not apply: incompatible changes may be introduced in any future release.
+	 */
+	SearchPredicateFactory predicate();
+
+	/**
+	 * @return attrs of the filter factory.
+	 */
+	Object[] attrs();
+
+	/**
+	 * @param <T> type of the parameters
+	 * @param name a name of parametr
+	 * @return parametr of the filter factory {@link FilterPredicateOptionsStep#param(java.lang.String, java.lang.Object)}.
+	 */
+	<T> T param(String name);
+
+	/**
+	 * @return parent path.
+	 */
+	String parentPath();
+
+	/**
+	 * @return nested path.
+	 */
+	String nestedPath();
+
+	/**
+	 * @return absolute path.
+	 */
+	String absolutePath();
```

Comment:
Ideally these methods shouldn't be necessary.

It should be doable to remove the need for `parentPath()`, `absolutePath()` and `resolvePath(String)` if we make the `SearchPredicateFactory` returned by `predicate()` "scoped" to the field path, so that all paths passed to that factory are automatically prefixed with `parentPath`. Then you wouldn't need to worry about the `parentPath` in the `FilterFactoryContext`. Please don't work on that though, I can do it in a later PR (I'll have to think about the many tests it requires).

As to `nestedPath()`... As far as I can tell, the predicate returned by the factory will automatically be wrapped in a `nested` predicate if the parent fields are nested. See `org.hibernate.search.backend.lucene.search.predicate.impl.AbstractLuceneNestablePredicate#toQuery`. So I don't think exposing `nestedPath()` is necessary? If that's right, please remove it.

Reply:
Additional information is useful when building native queries. 

```java
result = predicate.extension(LuceneExtension.get())
                    .fromLuceneQuery(query)
                    .toPredicate();
```

```java
private static class PerrmissionSearchPredicate implements NamedPredicateFactory {

        @Override
        public SearchPredicate create(NamedPredicateFactoryContext ctx) {
            SearchPredicate result;
            String fieldPath = ctx.resolvePath((String) ctx.param("field"));
            Account user = (Account) ctx.param("user");
            String operation = (String) ctx.param("operation");

            if (operation == null) {
                operation = "read";
            }

            SearchPredicateFactory predicate = ctx.predicate();

            PermissionQuery allowQuery = new PermissionQuery(fieldPath, operation, user, true);
            PermissionQuery dannyQuery = new PermissionQuery(fieldPath, operation, user, false);

            BooleanQuery query = new BooleanQuery.Builder()
                    .add(allowQuery, BooleanClause.Occur.MUST)
                    .add(dannyQuery, BooleanClause.Occur.MUST_NOT)
                    .build();

            result = predicate.extension(LuceneExtension.get())
                    .fromLuceneQuery(query)
                    .toPredicate();

            return result;
        }
    }
```
I had to provide an absolute path to build my own native query in "LuceneQuery". The rest of the functions follow from the analysis that they can be useful in such cases.

Reply:
Here just a total use case, what the result of my need to extract, sort fields by permission to the content of the entity. There can be many such cases, so it is safe to leave access to such information to programmers. 

```java
public class PermissionsBinder implements TypeBinder {

    private String fieldName = "permissions";
    private String operationName = "read";
    private String filterName = "has-permission";

    public PermissionsBinder fieldName(String fieldName) {
        this.fieldName = fieldName;
        return this;
    }

    public PermissionsBinder operationName(String operationName) {
        this.operationName = operationName;
        return this;
    }

    public PermissionsBinder filterName(String filterName) {
        this.filterName = filterName;
        return this;
    }

    @Override
    public void bind(TypeBindingContext context) {
        context.dependencies().useRootOnly();

        IndexFieldReference<String> permissionsField = context.indexSchemaElement()
                .field(this.fieldName, f -> f.asString()
                .aggregable(Aggregable.NO)
                .sortable(Sortable.NO)
                .searchable(Searchable.YES)
                .analyzer("permissions"))
                .toReference();

        context.indexSchemaElement().namedPredicate(this.filterName, new PerrmissionSearchPredicate())
                .param("operation", operationName)
                .param("field", this.fieldName)
                .toReference();

        context.bridge(new Bridge(
                permissionsField
        ));
    }

    private static class Bridge implements TypeBridge {

        private final IndexFieldReference<String> permissionsField;

        protected IdentityStoreResolver getIdentityService() {
            IdentityStoreResolver isr = ServiceLocator.getService(IdentityStoreResolver.class, IdentityStoreResolver.JNDI_NAME);
            return isr;
        }

        private Bridge(IndexFieldReference<String> permissionsField) {
            this.permissionsField = permissionsField;
        }

        @Override
        public void write(DocumentElement target, Object bridgedElement, TypeBridgeWriteContext context) {
            StringBuilder brules = new StringBuilder();

            IdentityStoreResolver isr = getIdentityService();
            PermissionManager pm = isr.getPermissionManager();

            List<Permission> permissions = pm.listPermissions(bridgedElement);
            for (Permission permission : permissions) {
                IdentityType assignee = permission.getAssignee();

                if (assignee instanceof Group) {
                    String[] roles = permission.getRoles();
                    if (roles == null || roles.length < 1) {
                        roles = new String[]{"*"};
                    }
                    for (String role : roles) {
                        brules.append(permission.isAllow() ? "+" : "-");
                        brules.append(permission.getOperation()).append(":");
                        brules.append(permission.getContext()).append(":");
                        brules.append(role).append(":");
                        brules.append(permission.getPath());
                        brules.append("\n");
                    }
                } else {
                    brules.append(permission.isAllow() ? "+" : "-");
                    brules.append(permission.getOperation()).append(":");
                    brules.append(permission.getContext()).append(":");
                    brules.append("*").append(":");
                    brules.append(permission.getPath());
                    brules.append("\n");
                }
            }

            String rules = brules.toString();
            target.addValue(this.permissionsField, rules);
        }
    }

    private static class PerrmissionSearchPredicate implements NamedPredicateFactory {

        @Override
        public SearchPredicate create(NamedPredicateFactoryContext ctx) {
            SearchPredicate result;
            String fieldPath = ctx.resolvePath((String) ctx.param("field"));
            Account user = (Account) ctx.param("user");
            String operation = (String) ctx.param("operation");

            if (operation == null) {
                operation = "read";
            }

            SearchPredicateFactory predicate = ctx.predicate();

            PermissionQuery allowQuery = new PermissionQuery(fieldPath, operation, user, true);
            PermissionQuery dannyQuery = new PermissionQuery(fieldPath, operation, user, false);

            BooleanQuery query = new BooleanQuery.Builder()
                    .add(allowQuery, BooleanClause.Occur.MUST)
                    .add(dannyQuery, BooleanClause.Occur.MUST_NOT)
                    .build();

            result = predicate.extension(LuceneExtension.get())
                    .fromLuceneQuery(query)
                    .toPredicate();

            return result;
        }

    }
}
```

Reply:
You're right, it's useful for native queries.

Please bear with me, I'm trying to expose only what's necessary, because exposing unnecessary methods frequently paints us into a corner when we try to implement a new feature that would be incompatible with those unnecessary methods. It's hard to predict what will and what won't be a problem.

Another reason I'd rather make path resolution "implicit" is to avoid mistakes where people forget to resolve paths. Granted that's not possible for native predicates, but we can still address the easy use cases. I created [HSEARCH-4162](https://hibernate.atlassian.net/browse/HSEARCH-4162) to address that later.

So I'll try another question... from what I can see, only `.resolvePath(...)` was used in your code, so we can probably remove `absolutePath()`, `nestedPath()`, `parentPath()`? I can't think of a use case where we'd need these for anything other than resolving a path.

Reply:
As for "resolvePath ()", this is necessary for native queries. For the rest, it's hard to predict what a programmer would like to achieve. Probably "parentPath ()" and "absolutePath ()" inform about where to use predicate being built. I provided "nestedPath ()" prophylactically, assuming that they might be useful. They're on hand when building a context, so it doesn't make code anyway. I can hide them in the API, and if there is a use case, it will be revealed on request.

Reply:
Yes please remove these methods and we'll add them if someone asks for them.